### PR TITLE
[Slot] Access ref from props else fallback to element.ref

### LINKED ISSUE #2769: [Slot] React 18.3.0 canary, ref is now a regular prop
## Bug report

### Current Behavior

Next.js [14.1.1-canary.74](https://github.com/vercel/next.js/releases/tag/v14.1.1-canary.74) is pulling in the latest React 18.3.0 canary release has this upstream change: [Pass ref as normal prop \(#28348\)](https://github.com/facebook/react/pull/28348). This is throwing a warning: `Warning: Accessing element.ref is no longer supported. ref is now a regular prop. It will be removed from the JSX Element type in a future release.`

I assume it is the following line of code that is causing the warning:

https://github.com/radix-ui/primitives/blob/be80c2a309c1afd07f4176c742ea16ac0675659f/packages/react/slot/src/Slot.tsx#L66

**Full error message**

```
Warning: Accessing element.ref is no longer supported. ref is now a regular prop. It will be removed from the JSX Element type in a future release.   
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-slot@1.0.2_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-slot/dist/index.mjs:46:23)
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-slot@1.0.2_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-slot/dist/index.mjs:20:23)                                                                                                                                 
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-primitive@1.0.3_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-primitive/dist/index.mjs:44:26)                                                                             
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-popper@1.1.3_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-popper/dist/index.mjs:80:28)                                                                                   
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-menu@2.0.6_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-menu/dist/index.mjs:211:26)                                                                                      
    at eval (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-dropdown-menu@2.0.6_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-dropdown-menu/dist/index.mjs:110:34)                                                                    
    at Provider (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-context@1.0.1_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-context/dist/index.mjs:47:28)                                                                                                                       
    at Provider (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-context@1.0.1_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-context/dist/index.mjs:47:28)                                                                                                                       
    at Provider (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-context@1.0.1_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-context/dist/index.mjs:47:28)                                                                                                                       
    at $cf1ac5d9fe0e8206$export$badac9ada3a0bdf9 (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-popper@1.1.3_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-popper/dist/index.mjs:65:28)                                              
    at $6cc32821e9371a1c$export$d9b273488cd8ce6f (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-menu@2.0.6_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-menu/dist/index.mjs:150:26)
    at Provider (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-context@1.0.1_@types+react@18.2.54_react@18.2.0/node_modules/@radix-ui/react-context/dist/index.mjs:47:28)                                                                                                                       
    at $d08ef79370b62062$export$e44a253a59704894 (webpack-internal:///(ssr)/../../node_modules/.pnpm/@radix-ui+react-dropdown-menu@2.0.6_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0/node_modules/@radix-ui/react-dropdown-menu/dist/index.mjs:75:34)
    at DarkModeToggle (webpack-internal:///(ssr)/./src/components/layout/status/DarkModeToggle.tsx:19:13)                                                                                                                                                                                                            
    at Lazy                                                                                                                                                                                                                                                                                                          
    at div                                                                                                                                                                                                                                                                                                           
    at div                                                                                                                                                                                                                                                                                                           
    at div                                                                                                                                                                                                                                                                                                           
    at div                                                            
    at MenuProvider (webpack-internal:///(ssr)/./src/components/providers/MenuProvider.tsx:15:25)                                                     
    at Lazy                                                           
    at ApiClientProvider (webpack-internal:///(ssr)/../../packages/core/src/providers/ApiClientProvider.tsx:23:34)                                    
    at ApiClientProvider (webpack-internal:///(ssr)/./src/features/api/ApiClientProvider.tsx:15:30)                                                   
    at QueryClientProvider (webpack-internal:///(ssr)/../../node_modules/.pnpm/@tanstack+react-query@5.17.19_react@18.2.0/node_modules/@tanstack/react-query/build/modern/QueryClientProvider.js:21:30)
    at Provider (webpack-internal:///(ssr)/../../node_modules/.pnpm/react-redux@8.1.3_@types+react-dom@18.2.14_@types+react@18.2.54_react-dom@18.2.0_react@18.2.0_redux@4.2.1/node_modules/react-redux/es/components/Provider.js:16:3)
    at Providers (webpack-internal:///(ssr)/./src/components/providers/Providers.tsx:26:22)                                                           
    at Lazy                                                           
    at WebsocketProvider (webpack-internal:///(ssr)/../../packages/websockets/src/WebsocketProvider.tsx:20:30)                                        
    at Lazy                                                           
    at InnerLayoutRouter (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/layout-router.js:241:11)
    at RedirectErrorBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/redirect-boundary.js:72:9)
    at RedirectBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/redirect-boundary.js:80:11)
    at NotFoundErrorBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/not-found-boundary.js:76:9)
    at NotFoundBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/not-found-boundary.js:84:11)
    at LoadingBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/layout-router.js:341:11)
    at ErrorBoundaryHandler (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/error-boundary.js:113:9)
    at ErrorBoundary (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/error-boundary.js:159:11)
    at InnerScrollAndFocusHandler (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/layout-router.js:152:9)
    at ScrollAndFocusHandler (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/layout-router.js:227:11)
    at RenderFromTemplateContext (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/render-from-template-context.js:16:44)
    at Lazy                                                           
    at OuterLayoutRouter (webpack-internal:///(ssr)/../../node_modules/.pnpm/next@14.2.0-canary.9_@babel+core@7.23.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/client/components/layout-router.js:359:11)
    at Lazy                                                           
    at f (webpack-internal:///(ssr)/../../node_modules/.pnpm/next-themes@0.2.1_next@14.2.0-canary.9_react-dom@18.2.0_react@18.2.0/node_modules/next-themes/dist/index.module.js:8:597)
    at $ (webpack-internal:///(ssr)/../../node_modules/.pnpm/next-themes@0.2.1_next@14.2.0-canary.9_react-dom@18.2.0_react@18.2.0/node_modules/next-themes/dist/index.module.js:8:348)
    at ThemeProvider (webpack-internal:///(ssr)/./src/components/providers/ThemeProvider.tsx:10:26)
```

### Expected behaviour

No warning

### Reproducible example

*Will update with a repro*

### Suggested solution

n/a

### Additional context

I am using Radix via shadcn-ui

### Your environment

| Software         | Name(s) | Version                          |
| ---------------- | ------- | -------------------------------- |
| Radix Package(s) | slot    | 1.0.2                            |
| React            | n/a     | 18.3.0-canary-6c3b8dbfe-20240226 |
| Browser          |         | Firefox                          |
| Assistive tech   | n/a     |                                  |
| Node             | n/a     |                                  |
| npm/yarn         | pnpm    | 8.15.4                           |
| Operating System | Linux   |                                  |

### PR DESCRIPTION
**Pull request by chungweileong94**
### Description

fix #2769

Starting from the React 18.3.0 canary, accessing the ref via `element.ref` will throw a warning, and `element.ref` will be removed in the future React release.

Instead, we should access the ref from props. To maintain the backward compatibility, we can do `children.props.ref ?? children.ref`.

### CODE REVIEW

**Review by dgriffith2023 — APPROVED** [review]

**Comment by ZipBrandon** [comment]
Just chiming in to say that I had an issue with Portal when used in conjunction with `framer-motion` in the React canary releases since [19.0.0-canary-7a2609eed-20240403](https://www.npmjs.com/package/react/v/19.0.0-canary-7a2609eed-20240403) and that this was the fix. Thank you.

**Comment by alaney2** [comment]
Thanks for fixing this

**Comment by borispoehland** [comment]
Thank you!

**Comment by tillka** [comment]
Is there a way to fix this manually / locally until the official fix is out? The 7x 'element.ref' warnings I am getting on each page refresh are very annoying.

**Comment by juliusmarminge** [comment]
I also had to patch `@radix-ui/react-presence` to get rid of all warnings:

CleanShot 2024-04-26 at 21 31 55@2x

**Comment by chungweileong94** [comment]
> I also had to patch `@radix-ui/react-presence` to get rid of all warnings:
> 
> CleanShot 2024-04-26 at 21 31 55@2x

Ah thx, wasn’t able to find the others, will include these as well.

**Comment by chungweileong94** [comment]
@juliusmarminge I applied the same fix to `@radix-ui/react-presence` too.

**Review by juliusmarminge — APPROVED** [review]
LGTM

File: packages/react/presence/src/Presence.tsx
Comment (eps1lon) [inline packages/react/presence/src/Presence.tsx:24]: It's fine to access `child.ref` conditionally. React only warns when accessing `child.ref` iff a `ref` prop is actually passed.

**Review by nicksrandall — APPROVED** [review]
LGTM

**Comment by Brenndoerfer** [comment]
Would be great to get an approval from one of the code owners on this.. it also impacts Shadcn/ui

**Review by Brenndoerfer — APPROVED** [review]

**Review by gtjamesa — APPROVED** [review]

**Comment by paplco** [comment]
hey @andy-hook @benoitgrelard 

hoping this gets approved soon. thanks!

**Comment by sannajammeh** [comment]
As this is taking a little while, here's how to suppress the annoying logs until its merged:

`_supressLogs.ts`
```tsx 
const prevConsoleError = console.error;
const prevConsoleWarn = console.warn;

console.error = (...args) => {
  if (args[0].includes("Warning: Accessing element.ref")) {
    return;
  }

  prevConsoleError(...args);
};

console.warn = (...args) => {
  if (args[0].includes("Warning: Accessing element.ref")) {
    return;
  }

  prevConsoleWarn(...args);
};
```

Import this globally at your app root.
For Next.js this would be

Root `layout.tsx`
```ts
import "./_supressLogs";
// Your root layout
```

And Root client provider file `_Providers.tsx`
```ts
"use client";

import "./_supressLogs";
// Your provider
```

**Review by ferothefox — APPROVED** [review]

**Review by royce-mathew — APPROVED** [review]

**Comment by bcanfield** [comment]
lgtm

**Review by xuanhao-bic — APPROVED** [review]

**Review by mooxl — APPROVED** [review]

**Comment by mooxl** [comment]
3.1.0 is out and resolves the `ref` issue 🪩

**Comment by louishugens** [comment]
Hello guys. I still get this error: Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release. Is 1.1.0 the last version? I have a nextjs 15 project

**Comment by chungweileong94** [comment]
> Hello guys. I still get this error: Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release. Is 1.1.0 the last version? I have a nextjs 15 project

That should be the latest version that contains the fix. If you are using component from [shadcn ui](https://github.com/shadcn-ui/ui), you will need to manually remove all the `forwardRef` usages. But if you can provide a small repro if that doesn't solve your issue.

**Comment by louishugens** [comment]
Thanks @chungweileong94. I am indeed using shadcn-ui. It is complaining about this file node_modules/@radix-ui/react-menu/node_modules/@radix-ui/react-slot/dist/packages/react/slot/src/Slot.tsx (66:71) but i can't locate it

**Comment by chungweileong94** [comment]
> Thanks @chungweileong94. I am indeed using shadcn-ui. It is complaining about this file node_modules/@radix-ui/react-menu/node_modules/@radix-ui/react-slot/dist/packages/react/slot/src/Slot.tsx (66:71) but i can't locate it

I looks like a dependency problem to me, have you try to update all the radix packages to the latest version?

**Comment by louishugens** [comment]
i did try to update. No success. Also i get this from time to time: `Cannot update a component (`HotReload`) while rendering a different component (`SlotClone`). To locate the bad setState() call inside `SlotClone`, follow the stack trace as described in https://react.dev/link/setstate-in-render`

**Comment by lpbonomi** [comment]
@louishugens @chungweileong94 I think that the problem is the `react-slot` package still has `forwardedRef` in it:

https://github.com/radix-ui/primitives/blob/main/packages/react/slot/src/Slot.tsx

**Comment by chungweileong94** [comment]
> @louishugens @chungweileong94 I think that the problem is the `react-slot` package still has `forwardedRef` in it:
> 
> https://github.com/radix-ui/primitives/blob/main/packages/react/slot/src/Slot.tsx

It's been awhile since I last created this PR, but as far as I can remember, the warning that React was throwing isn't about `React.forwardRef`, but instead it was about accessing the ref via `element.ref`. But it could be something new that I wasn't aware of.

**Comment by f5io** [comment]
> @louishugens @chungweileong94 I think that the problem is the `react-slot` package still has `forwardedRef` in it:
> 
> https://github.com/radix-ui/primitives/blob/main/packages/react/slot/src/Slot.tsx

in an attempt to update to react 18.3, i am seeing exactly this issue. i was under the impression that `@radix-ui` is ready for react 19, apologies if i'm being dumb, but what am i missing here?

**Comment by chungweileong94** [comment]
> in an attempt to update to react 18.3, i am seeing exactly this issue. i was under the impression that `@radix-ui` is ready for react 19, apologies if i'm being dumb, but what am i missing here?

I don't think React 18.3 will suffer the same problem though. But mind if you can provide a small repro for the error that you are facing?

**Comment by f5io** [comment]
> > in an attempt to update to react 18.3, i am seeing exactly this issue. i was under the impression that `@radix-ui` is ready for react 19, apologies if i'm being dumb, but what am i missing here?
> 
> I don't think React 18.3 will suffer the same problem though. But mind if you can provide a small repro for the error that you are facing?

it appears after explicitly updating to `react@rc` this error disappears, as far as i can tell. if i find some time i will attempt a small repro against 18.3.

**Pull request closed without merging**
